app/services/voucher_adjustments_service.rb

```diff
@@ -65,6 +84,14 @@ def update_tax_adjustment_for(adjustment, tax_amount)
     tax_adjustment = @order.adjustments.find_or_initialize_by(adjustment_attributes)
     tax_adjustment.amount = tax_amount
     tax_adjustment.save
+
+    # Add metada so we know which voucher adjustment is Tax related
+    AdjustmentMetadata.create(
+      adjustment: tax_adjustment,
+      enterprise: adjustment.originator.enterprise,
+      fee_name: "Tax",
+      fee_type: "Voucher"
+    )
```

Comment:
Nice. I would probably build the data structure first and then call save only once. Rails may put it all in one transaction then.
```suggestion
    tax_adjustment.amount = tax_amount

    # Add metada so we know which voucher adjustment is Tax related
    tax_adjustment.metadata ||= AdjustmentMetadata.new(
      enterprise: adjustment.originator.enterprise,
      fee_name: "Tax",
      fee_type: "Voucher"
    )

    tax_adjustment.save
```

I also did an optional assignment here in case we are updating an existing adjustment which has metadata already. The metadata shouldn't change, I think, and therefore we don't need to update it.

Reply:
Nice one, That's correct the metadata doesn't need to change even if the adjustment is updated.